docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,147 @@
+
+Because ASO already uses the new APIs, adopting ASO in CAPZ accomplishes the goal of migrating off of the deprecated APIs. In addition, the declarative, Kubernetes-native interface exposed by ASO unlocks opportunities to simplify how CAPZ reconciles resources in Azure.
+
+### Goals
+
+- Adopt ASO in CAPZ to handle creating, updating, and deleting Azure resources
+- Determine a method to migrate CAPZ's individual Azure service interfaces one-by-one to use ASO
+- Determine how ASO will be installed on Cluster API management clusters
+
+### Non-Goals/Future Work
+
+- Entirely eliminate direct usage of any Azure Go SDK
+- Handle changes to Azure resources asynchronously
+- Allow users to interact with ASO resources directly, out-of-band with CAPZ
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+TODO
+
+### Requirements (Optional)
+
+#### Functional Requirements
+
+##### FR1
+
+TODO
+
+#### Non-Functional Requirements
+
+##### NFR1
+
+TODO
+
+### Implementation Details/Notes/Constraints
+
+- ASO needs to be installed on the management cluster
+  - Multi- or single-tenant?
+  - Install in capz-controller-manager `func main()`
+- GET requests are not as straightforward in the ASO model because BYO resources
+  may not initially exist in ASO
+- Current service pattern in CAPZ works for Azure services that are not
+  represented in ASO (resource health)
+- Each release of ASO can expose several Azure API versions of each resource,
+  including GA and preview API versions.
+
+### Security Model
+
+- ASO resources are subject to Kubernetes RBAC => capz-controller-manager will require many more permissions that may extend beyond what's minimally required.
+  - If a user already has their own ASO resources installed, CAPZ may require
+    permission to operate against them to handle the ASO resources it needs to
+    manage.
+
+### Risks and Mitigations
+
+- Introduction of functional regressions
+- Increased management cluster API server traffic
+- Conflicting user installations of ASO or ASO resources
+- Future breaking changes in ASO
+
+## Alternatives
+
+- Adopt new SDK directly
+  - Likely much more straightforward transition => can be done faster with lower risk
+
+## Upgrade Strategy
+
+- There will be at least one upgrade path between CAPZ versions where ASO resources don't exist in the old version and they're expected to exist in the new version
+  - Need a one-time bootstrapping process to create managed resources in ASO?
+- Once ASO is integrated, updates to CAPZ may include updates to the API versions used for ASO resources.
+  - API versions of ASO resources may also be periodically removed.
+
+## Additional Details
+
+### Test Plan
+
+- Leverage existing e2e tests
+- Add unit tests for new ASO integration
+- Run one-off tests against large clusters to catch performance regressions
```

Comment:
we should talk about telemetry somewhere. Currently we have traces and metrics for every SDK call made in CAPZ https://capz.sigs.k8s.io/developers/development.html#viewing-telemetry, if we move to ASO we will lose that. @mattchr does ASO currently emit traces/metrics for SDK calls?

Reply:
It looks like ASO exposes `azure_successful_requests_total`, `azure_failed_requests_total`, and `azure_requests_time_seconds` Prometheus metrics, but I don't see any OpenTelemetry integration.

Reply:
We don't have any OpenTelemetry integration currently. We have prom metrics for every SDK call made, but not traces. As I mentioned in my other comment this is something we'd be open to improving, although I'm not sure how we'd get distributed tracing to work through CRs (so that you could have a top-level trace that spanned N ASO resource creations for example)